doc/aaa/radius_authentication.md

```diff
@@ -0,0 +1,634 @@
+2. The PAM configuration causes an Access-Request to RADIUS server.
+3. RADIUS returns Access-Accept with Management-Privilege-Level (MPL)(Goto 4.),
+   or Access-Reject (Authentication Failed).
+4. Access-Accept: The MPL is cached.
+5. A getpwnam_r() call is made to obtain user information.
+6. If the RADIUS user is not found in the local /etc/passwd file, the NSS
+   RADIUS plugin uses the cached MPL information to return user information.
+7. The user has authenticated successfully.
+
+Note:
+1. If an application requests getpwnam_r() before any authentication is
+   done, the RADIUS NSS plugin can be configured to return a least privileged
+   (MPL 1) temporary user info. On subsequent sessions, the user info will
+   be returned per the cached MPL. SSH is an example of such an application.
+   Alternatively, the user can be asked to login to console for the first
+   time to cache the MPL. Subsequent sessions will now use the cached MPL.
+
+Following is the sequence of events during RADIUS authenticated login over
+SSH:
+
+```
+SSH Client              SONiC Device                     RADIUS Server
+                  SSH Server     NSS RADIUS Plugin
+----------------------------------------------------------------------
+
+    |                 |                 |                    |
+    |                 |                 |                    |
+    |  Login(1)       |                 |                    |
+    +---------------->|                 |                    |
+    |                 | getpwnam_r()(2) |                    |
+    |                 +---------------->|                    |
+    |                 |                 |                    |
+    |                 | user info.(3)   |                    |
+    |                 |<----------------+                    |
+    |                 |                 |                    |
+    |                 |                                      |
+    |                 |         Access-Request(4)            |
+    |                 +------------------------------------->|
+    |                 |                                      |
+    |                 |         Access-Accept(5)             |
+    |                 |<-------------------------------------+
+    |                 |                                      |
+    |             (6) |                 |                    |
+    |                 |                 |                    |
+    |  Success(7)     |                 |                    |
+    |<----------------+                 |                    |
+    |                 |                 |                    |
+```
+
+1. A user tries to login through a SSH client.
+2. A getpwnam_r() call is made to obtain user information.
+3. If the RADIUS user is not found in the local /etc/passwd file, the NSS
+   RADIUS plugin uses the cached MPL information to return user information.
+   If no MPL is available, and the many_to_one=a is configured, a least
```

Comment:
This is configured in the file as mentioned in the "NSS" section, 2nd paragraph. I'll add a pointer.
The primary advantage of using the many_to_one=a (anonymous) option  is that users do not have to be manually created beforehand by the admin. The primary downside is that unknown users (i.e. unknown before the first successful RADIUS login for that user) will appear as the least privileged user. I can add this information to the HLD as well.

The Management-Privilege-Level cache is only used by the RADIUS NSS plugin if the user is not present in the /etc/passwd file. The many-to_one=a (anonymous) option tells the RADIUS NSS plugin to use the configuration for the least privileged user in the User Privilege Table subsection, if there is no entry for the user in the Management-Privilege-Level cache.

Reply:
`If the RADIUS user is not found in the local /etc/passwd file, the NSS RADIUS plugin uses the cached MPL information to return user information. If no MPL is available, and the many_to_one=a is configured, a least privileged (MPL 1) temporary user info is returned. Otherwise, the login fails`

When would be the case, "user not found in /etc/passwd", but MPL cache is found ?

Reply:
This would be the many_to_one=y (map users) case. In that case, the RADIUS user is not present on the system, thus the RADIUS NSS plugin would need to return mapped user information based on the MPL cached information. This is also discussed in the following comment. I work on adding this information as well.

Reply:
Thanks

Reply:
Now, creating a temporary user is default, so there is no need of the "anonymous" option. The spec is updated.